File: src/main/java/com/example/pidev/service/user/PasswordResetService.java
package com.example.pidev.service.user;

import com.example.pidev.model.user.PasswordResetToken;
import com.example.pidev.utils.DBConnection;

import java.sql.*;
import java.time.LocalDateTime;
import java.util.UUID;

public class PasswordResetService {

    private Connection connection;
    private String currentToken;
    private int currentUserId;

    public PasswordResetService() {
        try {
            this.connection = new DBConnection().getConnection();
            System.out.println("✅ PasswordResetService: Connexion DB établie");
        } catch (Exception e) {
            System.err.println("❌ PasswordResetService: Erreur de connexion - " + e.getMessage());
            e.printStackTrace();
        }
    }

    // ========== GESTION DES TOKENS ==========

    /**
     * Crée un nouveau token de réinitialisation
     */
    public void createToken(PasswordResetToken token) {
        if (connection == null) {
            System.err.println("❌ createToken: Pas de connexion DB");
            return;
        }

        // Vérifier si la table existe, sinon la créer
        createTokenTableIfNotExists();

        // ✅ MODIFIER: Enlever 'created_at' de la requête
        String query = "INSERT INTO password_reset_tokens (token, user_id, expiry_date, used) " +
                "VALUES (?, ?, ?, FALSE)";  // Plus de created_at

        try (PreparedStatement stmt = connection.prepareStatement(query)) {
            stmt.setString(1, token.getToken());
            stmt.setInt(2, token.getUserId());
            stmt.setTimestamp(3, Timestamp.valueOf(token.getExpiryDate()));

            int rowsInserted = stmt.executeUpdate();

            if (rowsInserted > 0) {
                System.out.println("✅ Token créé avec succès pour l'utilisateur ID: " + token.getUserId());
                System.out.println("   Token: " + token.getToken());
                System.out.println("   Expire le: " + token.getExpiryDate());
            } else {
                System.err.println("❌ Échec de création du token");
            }

        } catch (SQLException e) {
            System.err.println("❌ Erreur createToken: " + e.getMessage());
            e.printStackTrace();
        }
    }

    /**
     * Crée la table des tokens si elle n'existe pas
     */
    private void createTokenTableIfNotExists() {
        String createTableQuery = "CREATE TABLE IF NOT EXISTS password_reset_tokens (" +
                "id INT AUTO_INCREMENT PRIMARY KEY, " +
                "token VARCHAR(255) NOT NULL UNIQUE, " +
                "user_id INT NOT NULL, " +
                "expiry_date DATETIME NOT NULL, " +
                "used BOOLEAN DEFAULT FALSE, " +
                "created_at DATETIME DEFAULT NOW(), " +
                "INDEX idx_token (token), " +
                "INDEX idx_user_id (user_id)" +
                ")";

        try (Statement stmt = connection.createStatement()) {
            stmt.execute(createTableQuery);
            System.out.println("✅ Table password_reset_tokens vérifiée/créée");
        } catch (SQLException e) {
            System.err.println("❌ Erreur création table: " + e.getMessage());
        }
    }

    /**
     * Valide un token
     */
    public boolean validateToken(String token) {
        if (connection == null) {
            System.err.println("❌ validateToken: Pas de connexion DB");
            return false;
        }

        if (token == null || token.trim().isEmpty()) {
            System.err.println("❌ validateToken: Token vide");
            return false;
        }

        String query = "SELECT user_id FROM password_reset_tokens " +
                "WHERE token = ? AND used = FALSE AND expiry_date > NOW()";

        try (PreparedStatement stmt = connection.prepareStatement(query)) {
            stmt.setString(1, token);

            ResultSet rs = stmt.executeQuery();

            if (rs.next()) {
                currentToken = token;
                currentUserId = rs.getInt("user_id");
                System.out.println("✅ Token valide pour l'utilisateur ID: " + currentUserId);
                return true;
            } else {
                System.out.println("❌ Token invalide ou expiré: " + token);
                return false;
            }

        } catch (SQLException e) {
            System.err.println("❌ Erreur validateToken: " + e.getMessage());
            e.printStackTrace();
            return false;
        }
    }

    /**
     * Marque un token comme utilisé
     */
    private void markTokenAsUsed(String token) {
        if (connection == null) return;

        String query = "UPDATE password_reset_tokens SET used = TRUE WHERE token = ?";

        try (PreparedStatement stmt = connection.prepareStatement(query)) {
            stmt.setString(1, token);
            int rowsUpdated = stmt.executeUpdate();

            if (rowsUpdated > 0) {
                System.out.println("✅ Token marqué comme utilisé: " + token);
            }
        } catch (SQLException e) {
            System.err.println("❌ Erreur markTokenAsUsed: " + e.getMessage());
        }
    }

    /**
     * Nettoie les tokens expirés
     */
    public void cleanExpiredTokens() {
        if (connection == null) return;

        String query = "DELETE FROM password_reset_tokens WHERE expiry_date < NOW() OR used = TRUE";

        try (Statement stmt = connection.createStatement()) {
            int rowsDeleted = stmt.executeUpdate(query);
            if (rowsDeleted > 0) {
                System.out.println("🧹 Nettoyage: " + rowsDeleted + " tokens expirés supprimés");
            }
        } catch (SQLException e) {
            System.err.println("❌ Erreur cleanExpiredTokens: " + e.getMessage());
        }
    }

    // ========== GESTION DES MOTS DE PASSE ==========

    /**
     * Réinitialise le mot de passe (SANS HASHAGE - en clair)
     * ⚠️ ATTENTION: Non sécurisé, à utiliser uniquement pour le développement
     */
    public boolean resetPassword(String newPassword) {
        if (connection == null) {
            System.err.println("❌ resetPassword: Pas de connexion DB");
            return false;
        }

        if (currentToken == null || currentUserId == 0) {
            System.err.println("❌ resetPassword: Aucun token valide en session");
            return false;
        }

        // Vérifier une dernière fois que le token est toujours valide
        if (!validateToken(currentToken)) {
            System.err.println("❌ resetPassword: Token expiré ou déjà utilisé");
            return false;
        }

        // ⚠️ STOCKAGE EN CLAIR - NON SÉCURISÉ
        String updateQuery = "UPDATE user_model SET password = ? WHERE Id_User = ?";

        try (PreparedStatement stmt = connection.prepareStatement(updateQuery)) {
            stmt.setString(1, newPassword);  // Stockage en clair !
            stmt.setInt(2, currentUserId);

            int rowsAffected = stmt.executeUpdate();

            if (rowsAffected > 0) {
                // Marquer le token comme utilisé
                markTokenAsUsed(currentToken);

                System.out.println("⚠️ ATTENTION: Mot de passe stocké EN CLAIR pour l'utilisateur ID: " + currentUserId);
                System.out.println("✅ Mot de passe réinitialisé avec succès");

                // Réinitialiser les variables de session
                currentToken = null;
                currentUserId = 0;

                return true;
            } else {
                System.err.println("❌ Aucun utilisateur trouvé avec l'ID: " + currentUserId);
                return false;
            }

        } catch (SQLException e) {
            System.err.println("❌ Erreur resetPassword: " + e.getMessage());
            e.printStackTrace();
            return false;
        }
    }

    /**
     * Version alternative avec hashage (recommandée pour la production)
     * Décommentez cette méthode si vous voulez utiliser le hashage
     */
    /*
    public boolean resetPasswordWithHash(String newPassword) {
        if (connection == null) return false;
        if (currentToken == null || currentUserId == 0) return false;

        if (!validateToken(currentToken)) return false;

        // Hasher le mot de passe
        String hashedPassword = hashPassword(newPassword);

        String updateQuery = "UPDATE user_model SET password = ? WHERE Id_User = ?";

        try (PreparedStatement stmt = connection.prepareStatement(updateQuery)) {
            stmt.setString(1, hashedPassword);
            stmt.setInt(2, currentUserId);

            int rowsAffected = stmt.executeUpdate();

            if (rowsAffected > 0) {
                markTokenAsUsed(currentToken);
                System.out.println("✅ Mot de passe hashé et réinitialisé pour l'utilisateur ID: " + currentUserId);
                currentToken = null;
                currentUserId = 0;
                return true;
            }

        } catch (SQLException e) {
            e.printStackTrace();
        }
        return false;
    }

    private String hashPassword(String password) {
        try {
            java.security.MessageDigest digest = java.security.MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(password.getBytes());
            return java.util.Base64.getEncoder().encodeToString(hash);
        } catch (java.security.NoSuchAlgorithmException e) {
            e.printStackTrace();
            return password;
        }
    }
    */

    // ========== MÉTHODES UTILITAIRES ==========

    /**
     * Génère un token unique
     */
    public String generateToken() {
        return UUID.randomUUID().toString();
    }

    /**
     * Récupère l'ID utilisateur associé à un token
     */
    public int getUserIdFromToken(String token) {
        if (connection == null) return -1;

        String query = "SELECT user_id FROM password_reset_tokens WHERE token = ?";

        try (PreparedStatement stmt = connection.prepareStatement(query)) {
            stmt.setString(1, token);
            ResultSet rs = stmt.executeQuery();

            if (rs.next()) {
                return rs.getInt("user_id");
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return -1;
    }

    /**
     * Vérifie si un token existe
     */
    public boolean tokenExists(String token) {
        if (connection == null) return false;

        String query = "SELECT COUNT(*) FROM password_reset_tokens WHERE token = ?";

        try (PreparedStatement stmt = connection.prepareStatement(query)) {
            stmt.setString(1, token);
            ResultSet rs = stmt.executeQuery();

            if (rs.next()) {
                return rs.getInt(1) > 0;
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return false;
    }

    /**
     * Supprime un token
     */
    public boolean deleteToken(String token) {
        if (connection == null) return false;

        String query = "DELETE FROM password_reset_tokens WHERE token = ?";

        try (PreparedStatement stmt = connection.prepareStatement(query)) {
            stmt.setString(1, token);
            int rowsDeleted = stmt.executeUpdate();
            return rowsDeleted > 0;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }

    /**
     * Récupère le nombre de tokens actifs pour un utilisateur
     */
    public int getActiveTokensCount(int userId) {
        if (connection == null) return 0;

        String query = "SELECT COUNT(*) FROM password_reset_tokens " +
                "WHERE user_id = ? AND used = FALSE AND expiry_date > NOW()";

        try (PreparedStatement stmt = connection.prepareStatement(query)) {
            stmt.setInt(1, userId);
            ResultSet rs = stmt.executeQuery();

            if (rs.next()) {
                return rs.getInt(1);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return 0;
    }

    // ========== GETTERS/SETTERS ==========

    public String getCurrentToken() {
        return currentToken;
    }

    public void setCurrentToken(String token) {
        this.currentToken = token;
    }

    public int getCurrentUserId() {
        return currentUserId;
    }

    public void setCurrentUserId(int userId) {
        this.currentUserId = userId;
    }

    /**
     * Ferme la connexion
     */
    public void closeConnection() {
        try {
            if (connection != null && !connection.isClosed()) {
                connection.close();
                System.out.println("🔌 Connexion DB fermée");
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }
}

File: src/Service/User/PasswordResetService.php
<?php

namespace App\Service\User;

use App\Entity\User\UserModel;
use App\Entity\User\PasswordResetToken;

use App\Repository\User\PasswordResetTokenRepository;
use App\Repository\User\UserRepository;
use Doctrine\ORM\EntityManagerInterface;
use Psr\Log\LoggerInterface;
use Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface;
use Symfony\Component\Uid\Uuid;

class PasswordResetService
{
    private ?string $currentToken = null;
    private ?int $currentUserId = null;

    public function __construct(
        private EntityManagerInterface $entityManager,
        private PasswordResetTokenRepository $tokenRepository,
        private UserRepository $userRepository,
        private UserPasswordHasherInterface $passwordHasher,
        private LoggerInterface $logger,
        private EmailService $emailService
    ) {
        $this->logger->info('✅ PasswordResetService initialisé');
        
        // Nettoyer les tokens expirés au démarrage
        $this->cleanExpiredTokens();
    }

    // ========== GESTION DES TOKENS ==========

    /**
     * Crée un nouveau token de réinitialisation
     */
    public function createToken(UserModel $user): PasswordResetToken
    {
        try {
            // Désactiver les anciens tokens
            $this->deactivateExistingTokens($user);

            // Créer un nouveau token
            $token = new PasswordResetToken();
            $token->setToken(Uuid::v4()->toRfc4122());
            $token->setUser($user);
            $token->setExpiryDate(new \DateTime('+1 hour'));
            $token->setUsed(false);

            $this->entityManager->persist($token);
            $this->entityManager->flush();

            $this->logger->info('✅ Token créé pour l\'utilisateur ID: ' . $user->getId());
            $this->logger->debug('   Token: ' . $token->getToken());
            $this->logger->debug('   Expire le: ' . $token->getExpiryDate()->format('Y-m-d H:i:s'));

            return $token;

        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur createToken: ' . $e->getMessage());
            throw $e;
        }
    }

    /**
     * Désactive les tokens existants pour un utilisateur
     */
    private function deactivateExistingTokens(UserModel $user): void
    {
        $activeTokens = $this->tokenRepository->findActiveByUser($user->getId());
        
        foreach ($activeTokens as $token) {
            $token->setUsed(true);
            $this->entityManager->persist($token);
        }
        
        $this->entityManager->flush();
        
        if (count($activeTokens) > 0) {
            $this->logger->info('🔄 ' . count($activeTokens) . ' anciens tokens désactivés');
        }
    }

    /**
     * Valide un token
     */
    public function validateToken(string $token): bool
    {
        try {
            if (empty($token)) {
                $this->logger->warning('❌ validateToken: Token vide');
                return false;
            }

            $resetToken = $this->tokenRepository->findValidToken($token);

            if ($resetToken) {
                $this->currentToken = $token;
                $this->currentUserId = $resetToken->getUser()->getId();
                
                $this->logger->info('✅ Token valide pour l\'utilisateur ID: ' . $this->currentUserId);
                return true;
            }

            $this->logger->warning('❌ Token invalide ou expiré: ' . $token);
            return false;

        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur validateToken: ' . $e->getMessage());
            return false;
        }
    }

    /**
     * Récupère l'entité token à partir du token string
     */
    private function getTokenEntity(string $token): ?PasswordResetToken
    {
        return $this->tokenRepository->findOneBy(['token' => $token]);
    }

    /**
     * Marque un token comme utilisé
     */
    private function markTokenAsUsed(string $token): bool
    {
        try {
            $resetToken = $this->getTokenEntity($token);
            
            if (!$resetToken) {
                $this->logger->warning('❌ Token non trouvé: ' . $token);
                return false;
            }

            $resetToken->setUsed(true);
            $this->entityManager->flush();

            $this->logger->info('✅ Token marqué comme utilisé: ' . $token);
            return true;

        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur markTokenAsUsed: ' . $e->getMessage());
            return false;
        }
    }

    /**
     * Nettoie les tokens expirés
     */
    public function cleanExpiredTokens(): int
    {
        try {
            $expiredTokens = $this->tokenRepository->findExpiredTokens();
            
            foreach ($expiredTokens as $token) {
                $this->entityManager->remove($token);
            }
            
            $this->entityManager->flush();

            $count = count($expiredTokens);
            if ($count > 0) {
                $this->logger->info('🧹 Nettoyage: ' . $count . ' tokens expirés supprimés');
            }

            return $count;

        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur cleanExpiredTokens: ' . $e->getMessage());
            return 0;
        }
    }

    // ========== GESTION DES MOTS DE PASSE ==========

    /**
     * Réinitialise le mot de passe (version sécurisée avec hashage)
     */
    public function resetPassword(string $newPassword): bool
    {
        try {
            if (!$this->currentToken || !$this->currentUserId) {
                $this->logger->error('❌ resetPassword: Aucun token valide en session');
                return false;
            }

            // Vérifier une dernière fois que le token est valide
            if (!$this->validateToken($this->currentToken)) {
                $this->logger->error('❌ resetPassword: Token expiré ou déjà utilisé');
                return false;
            }

            // Récupérer l'utilisateur
            $user = $this->userRepository->find($this->currentUserId);
            
            if (!$user) {
                $this->logger->error('❌ resetPassword: Utilisateur non trouvé ID: ' . $this->currentUserId);
                return false;
            }

            // Hasher le nouveau mot de passe (sécurisé)
            $hashedPassword = $this->passwordHasher->hashPassword($user, $newPassword);
            $user->setPassword($hashedPassword);

            // Marquer le token comme utilisé
            $this->markTokenAsUsed($this->currentToken);

            $this->entityManager->flush();

            $this->logger->info('✅ Mot de passe réinitialisé pour l\'utilisateur ID: ' . $this->currentUserId);
            
            // Réinitialiser les variables de session
            $this->currentToken = null;
            $this->currentUserId = null;

            return true;

        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur resetPassword: ' . $e->getMessage());
            return false;
        }
    }

    /**
     * Version NON sécurisée pour compatibilité avec JavaFX (mots de passe en clair)
     * ⚠️ À UTILISER UNIQUEMENT POUR LA MIGRATION
     */
    public function resetPasswordInsecure(string $newPassword): bool
    {
        try {
            if (!$this->currentToken || !$this->currentUserId) {
                $this->logger->error('❌ resetPasswordInsecure: Aucun token valide');
                return false;
            }

            if (!$this->validateToken($this->currentToken)) {
                return false;
            }

            $user = $this->userRepository->find($this->currentUserId);
            
            if (!$user) {
                return false;
            }

            // ⚠️ Stockage en clair (NON SÉCURISÉ - pour migration uniquement)
            $user->setPassword($newPassword); 

            $this->markTokenAsUsed($this->currentToken);
            $this->entityManager->flush();

            $this->logger->warning('⚠️ ATTENTION: Mot de passe stocké EN CLAIR pour l\'utilisateur ID: ' . $this->currentUserId);
            
            $this->currentToken = null;
            $this->currentUserId = null;

            return true;

        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur resetPasswordInsecure: ' . $e->getMessage());
            return false;
        }
    }

    // ========== MÉTHODES UTILITAIRES ==========

    /**
     * Génère un token unique
     */
    public function generateToken(): string
    {
        return Uuid::v4()->toRfc4122();
    }

    /**
     * Récupère l'ID utilisateur associé à un token
     */
    public function getUserIdFromToken(string $token): ?int
    {
        try {
            $resetToken = $this->tokenRepository->findOneBy(['token' => $token]);
            return $resetToken?->getUser()->getId();
        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur getUserIdFromToken: ' . $e->getMessage());
            return null;
        }
    }

    /**
     * Vérifie si un token existe
     */
    public function tokenExists(string $token): bool
    {
        try {
            return $this->tokenRepository->count(['token' => $token]) > 0;
        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur tokenExists: ' . $e->getMessage());
            return false;
        }
    }

    /**
     * Supprime un token
     */
    public function deleteToken(string $token): bool
    {
        try {
            $resetToken = $this->getTokenEntity($token);
            
            if (!$resetToken) {
                return false;
            }

            $this->entityManager->remove($resetToken);
            $this->entityManager->flush();

            $this->logger->info('✅ Token supprimé: ' . $token);
            return true;

        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur deleteToken: ' . $e->getMessage());
            return false;
        }
    }

    /**
     * Récupère le nombre de tokens actifs pour un utilisateur
     */
    public function getActiveTokensCount(int $userId): int
    {
        try {
            return count($this->tokenRepository->findActiveByUser($userId));
        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur getActiveTokensCount: ' . $e->getMessage());
            return 0;
        }
    }

    // ========== PROCESSUS COMPLET DE RÉINITIALISATION ==========

    /**
     * Processus complet de demande de réinitialisation
     */
    public function requestPasswordReset(string $email): bool
    {
        try {
            $user = $this->userRepository->findOneBy(['email' => $email]);

            if (!$user) {
                // Ne pas révéler si l'email existe ou pas (sécurité)
                $this->logger->info('📧 Demande de reset pour email non trouvé: ' . $email);
                return true;
            }

            // Créer un token
            $token = $this->createToken($user);

            // Envoyer l'email
            $this->emailService->sendPasswordResetEmail($user, $token->getToken());

            $this->logger->info('📧 Email de réinitialisation envoyé à: ' . $email);
            return true;

        } catch (\Exception $e) {
            $this->logger->error('❌ Erreur requestPasswordReset: ' . $e->getMessage());
            return false;
        }
    }

    // ========== GETTERS/SETTERS ==========

    public function getCurrentToken(): ?string
    {
        return $this->currentToken;
    }

    public function setCurrentToken(?string $token): self
    {
        $this->currentToken = $token;
        return $this;
    }

    public function getCurrentUserId(): ?int
    {
        return $this->currentUserId;
    }

    public function setCurrentUserId(?int $userId): self
    {
        $this->currentUserId = $userId;
        return $this;
    }
}